cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -404,6 +404,71 @@ struct CastFunctor<
   }
 };
 
+// Decimal to Integer
+
+template <typename O, typename I, typename Enable = void>
+struct is_decimal_truncate {
+  static constexpr bool value = false;
+};
+
+template <typename O, typename I>
+struct is_decimal_truncate<
+    O, I, enable_if_t<(is_integer_type<O>::value && is_decimal_type<I>::value)>> {
+  static constexpr bool value = true;
+};
+
+template <typename O, typename I>
+struct CastFunctor<O, I, enable_if_t<is_decimal_truncate<O, I>::value>> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    using in_type = typename I::c_type;
+    using out_type = typename O::c_type;
+    const auto& in_type_inst = checked_cast<const I&>(*input.type);
+    auto in_scale = in_type_inst.scale();
+
+    const in_type* in_data = input.GetValues<in_type>(1);
+    auto out_data = output->GetMutableValues<out_type>(1);
+
+    for (int64_t i = 0; i < input.length; ++i) {
+      auto result = in_data++->Rescale(in_scale, 0);
+      if (result.ok()) {
+        *out_data++ = static_cast<out_type>(result.ValueOrDie());
+      } else {
+        ARROW_LOG(ERROR) << result.status();
+      }
+    }
+  }
+};
+
+// Decimal to Decimal
+
+template <typename O, typename I>
+struct CastFunctor<
+    O, I, enable_if_t<(is_decimal_type<O>::value && is_decimal_type<I>::value)>> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    using in_type = typename I::c_type;
+    using out_type = typename O::c_type;
+
+    const auto& in_type_inst = checked_cast<const I&>(*input.type);
+    const auto& out_type_inst = checked_cast<const O&>(*output->type);
+    auto in_scale = in_type_inst.scale();
+    auto out_scale = out_type_inst.scale();
+
+    const in_type* in_data = input.GetValues<in_type>(1);
+    auto out_data = output->GetMutableValues<out_type>(1);
+
+    for (int64_t i = 0; i < input.length; ++i) {
+      auto result = in_data++->Rescale(in_scale, out_scale);
+      if (result.ok()) {
+        *out_data++ = static_cast<out_type>(result.ValueOrDie());
+      } else {
+        ARROW_LOG(ERROR) << result.status();
```

Comment:
Same here: should save the error on the context.

Reply:
Corrected.